### DISCUSSION TITLE
Why not compare two addresses directly?

### DISCUSSION THREAD

**Discussion by lindianfeng** [transmission/transmission#5679] [General]
int tr_address::compare(tr_address const& that) const noexcept // <=>
{
    // IPv6 addresses are always "greater than" IPv4
    if (this->type != that.type)
    {
        return this->is_ipv4() ? 1 : -1;
    }

    // in_addr_t is by definition uint32_t, so we convert it to int64_t,
    // the smallest signed integer type that can contain all values of uint32_t.
    return this->is_ipv4() ? static_cast<int>(int64_t{ ntohl(this->addr.addr4.s_addr) } - ntohl(that.addr.addr4.s_addr)) :
                             memcmp(&this->addr.addr6.s6_addr, &that.addr.addr6.s6_addr, sizeof(this->addr.addr6.s6_addr));
}

why the cmpare the ipv4 is not  this->addr.addr4.s_addr == that.addr.addr4.s_addr ？

**Comment by ckerr** [comment]
It's not an equality comparison, it's a three-way compare. Since Transmission is C++17 we don't have the `<=>` operator but the idea is the same, we want to return <0 if a < b, 0 if a == b, and >0 if a > b.

**Reply by lindianfeng** [reply]
get it
